### PR TITLE
Pass to rewrite Numpy function names to be able to overload them for Numba-dppy pipeline

### PR DESCRIPTION
**Pull request by reazulhoque**
The PR introduces a pass to rewrite the name of the Numpy function we intend to overload for Numba-dppy.

1. We rewrite the name of selected functions to `numba_dppy.dpnp.function_name`.
2. We create stub entried to make sure Numba finds the rewritten functions in `numab_dppy.dpnp` module.
3. We provide implementation of the functions using the `@overload` decorator.

Currnetly it has examples of how we can use this method to overload `np.sum` and `np.linalg.eig` (incomplete). We will need to port existing functions we provide support for and also new ones.

### CODE REVIEW

File: numba_dppy/device_init.py [hunk 1]
```diff
@@ -18,6 +18,10 @@
     CLK_GLOBAL_MEM_FENCE,
 )
 
+from .dpnp_glue.stubs import (
```
Comment (diptorupd) [inline numba_dppy/device_init.py:21]: A comment will be helpful.

File: numba_dppy/dpnp_glue/dpnpdecl.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+from numba.core.typing.templates import (AttributeTemplate, infer, infer_getattr, AbstractTemplate, signature)
+import numba_dppy
+from numba import types
+
+@infer_getattr
+class DpplDpnpTemplate(AttributeTemplate):
```
Comment (diptorupd) [inline numba_dppy/dpnp_glue/dpnpdecl.py:6]: s/Dppl/Dppy/g we are migrating to new naming convention next release.

File: numba_dppy/dpnp_glue/dpnpimpl.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+class _DPNP_EXTENSION:
+    def __init__(self, name):
+        dpnp_lowering.ensure_dpnp(name)
+
+    @classmethod
+    def get_sycl_queue(cls):
+        ret_type  = types.voidptr
+        sig       = signature(ret_type)
+        return types.ExternalFunction("DPPLQueueMgr_GetCurrentQueue", sig)
```
Comment (diptorupd) [inline numba_dppy/dpnp_glue/dpnpimpl.py:43]: All these function names need to change. Look at https://github.com/IntelPython/dpctl/pull/201

File: numba_dppy/dppl_passbuilder.py [hunk 4]
```diff
@@ -44,6 +45,9 @@ def default_numba_nopython_pipeline(state, pm):
         pm.add_pass(IRProcessing, "processing IR")
         pm.add_pass(WithLifting, "Handle with contexts")
 
+        # this pass rewrite name of functions
+        pm.add_pass(DPPLRewriteOverloadedFunctions, "dppl rewrite name of functions")
```
Comment (diptorupd) [inline numba_dppy/dppl_passbuilder.py:49]: `dppy`. Also, the pass description needs to be improved.

File: numba_dppy/dppl_passes.py [hunk 5]
```diff
@@ -38,6 +40,80 @@ def dpnp_available():
...
+        import numba_dppy.dpnp_glue.dpnpimpl
+        for label in topo_order:
+            block = blocks[label]
+            saved_arr_arg = {}
+            new_body = []
+            for stmt in block.body:
+                if isinstance(stmt, ir.Assign) and isinstance(stmt.value, ir.Expr):
+                    lhs = stmt.target.name
+                    rhs = stmt.value
+                    # replace np.func with name from map np.func
```
Comment (diptorupd) [inline numba_dppy/dppl_passes.py:66]: Is there as typo in the comment?

File: numba_dppy/dppl_passes.py [hunk 6]
```diff
@@ -38,6 +40,80 @@ def dpnp_available():
...
+                            rhs = stmt.value
+                            rhs.attr = rewrite_function_name_map[rhs.attr][1]
+
+                            global_module = rhs.value
+                            saved_arr_arg[lhs] = global_module
+
+                            scope = global_module.scope
+                            loc = global_module.loc
+
+                            g_dppl_var = ir.Var(scope, mk_unique_var("$dppl_replaced_var"), loc)
```
Comment (diptorupd) [inline numba_dppy/dppl_passes.py:82]: s/dppl/dppy/g

File: numba_dppy/dppl_passes.py [hunk 7]
```diff
@@ -38,6 +40,80 @@ def dpnp_available():
...
         return False
 
+rewrite_function_name_map = {"sum": (["np"], "sum"),
+                             "eig": (["linalg"], "eig")}
+
+class RewriteOverloadedFunctions(object):
+    def __init__(self, state, rewrite_function_name_map=rewrite_function_name_map):
+        self.state = state
+        self.function_name_map = rewrite_function_name_map
+
```
Comment (diptorupd) [inline numba_dppy/dppl_passes.py:50]: A comment is needed showing an example of the result of the rewrite, _i.e._ before and after versions of the IR.

File: numba_dppy/dppl_passes.py [hunk 8]
```diff
@@ -38,6 +40,80 @@ def dpnp_available():
...
+            saved_arr_arg = {}
+            new_body = []
+            for stmt in block.body:
+                if isinstance(stmt, ir.Assign) and isinstance(stmt.value, ir.Expr):
+                    lhs = stmt.target.name
+                    rhs = stmt.value
+                    # replace np.func with name from map np.func
+                    if (rhs.op == 'getattr' and rhs.attr in rewrite_function_name_map):
+                        module_node = block.find_variable_assignment(rhs.value.name).value
+                        if ((isinstance(module_node, ir.Global) and
```
Comment (diptorupd) [inline numba_dppy/dppl_passes.py:69]: A comment explaining the logic is required. When do we need to rewrite a `Global` node?

File: numba_dppy/dppl_passes.py [hunk 9]
```diff
@@ -38,6 +40,80 @@ def dpnp_available():
     except:
         return False
 
+rewrite_function_name_map = {"sum": (["np"], "sum"),
+                             "eig": (["linalg"], "eig")}
+
+class RewriteOverloadedFunctions(object):
```
Comment (diptorupd) [inline numba_dppy/dppl_passes.py:46]: `s/RewriteOverloadedFunctions/RewriteNumPyOverloadedFunctions/g`

File: numba_dppy/dppl_passes.py [hunk 10]
```diff
@@ -38,6 +40,80 @@ def dpnp_available():
     except:
         return False
 
+rewrite_function_name_map = {"sum": (["np"], "sum"),
```
Comment (diptorupd) [inline numba_dppy/dppl_passes.py:43]: The new pass should be in its own file. And please format with `black`.

**Review by diptorupd — COMMENTED** [review]
@reazulhoque mostly high-level comments. Watch out for DPPL to DPPY renaming as that is coming in next release. We need documentation and comments.

File: numba_dppy/dppl_passes.py (hunk 10 above)
Comment (reazulhoque) [reply numba_dppy/dppl_passes.py]: Moved to new file.

File: numba_dppy/dppl_passes.py (hunk 9 above)
Comment (reazulhoque) [reply numba_dppy/dppl_passes.py]: Addressed.

File: numba_dppy/dppl_passes.py (hunk 8 above)
Comment (reazulhoque) [reply numba_dppy/dppl_passes.py]: Comment added.

File: numba_dppy/dppl_passes.py (hunk 7 above)
Comment (reazulhoque) [reply numba_dppy/dppl_passes.py]: Comment added.

File: numba_dppy/dppl_passes.py (hunk 6 above)
Comment (reazulhoque) [reply numba_dppy/dppl_passes.py]: Changed.

File: numba_dppy/dppl_passes.py (hunk 5 above)
Comment (reazulhoque) [reply numba_dppy/dppl_passes.py]: Comment fixed.

File: numba_dppy/dppl_passbuilder.py (hunk 4 above)
Comment (reazulhoque) [reply numba_dppy/dppl_passbuilder.py]: Desctiption updated. Please check.

File: numba_dppy/dpnp_glue/dpnpimpl.py (hunk 3 above)
Comment (reazulhoque) [reply numba_dppy/dpnp_glue/dpnpimpl.py]: Updated.

File: numba_dppy/dpnp_glue/dpnpdecl.py (hunk 2 above)
Comment (reazulhoque) [reply numba_dppy/dpnp_glue/dpnpdecl.py]: Updated.

File: numba_dppy/device_init.py (hunk 1 above)
Comment (reazulhoque) [reply numba_dppy/device_init.py:25]: Comment added.

File: numba_dppy/dpnp_glue/dpnpimpl.py [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+        copy_usm(sycl_queue, out.ctypes, out_usm, out.size * out.itemsize)
+
+        free_usm(a_usm, sycl_queue)
+        free_usm(out_usm, sycl_queue)
+
+        return out[0]
+
+    return dpnp_sum_impl
+
+'''
```
Comment (diptorupd) [inline numba_dppy/dpnp_glue/dpnpimpl.py:116]: I am not a fan of having commented out code in the source. Please remove this code and bring it back once ready as a fresh PR.

File: numba_dppy/dpnp_glue/dpnpimpl.py [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+    def dpnp_sum(cls, fn_name, type_names):
+        ret_type  = types.void
+        sig       = signature(ret_type, types.voidptr, types.voidptr, types.int64)
+        f_ptr     = get_dpnp_fptr(fn_name, type_names)
+        def get_pointer(obj):
+            return f_ptr
+        return types.ExternalFunctionPointer(sig, get_pointer=get_pointer)
+
+    @classmethod
+    def dpnp_eig(cls, fn_name, type_names):
```
Comment (diptorupd) [inline numba_dppy/dpnp_glue/dpnpimpl.py:74]: Since the overload is not yet ready, we should remove this function from this PR and bring it back in a follow up.

File: numba_dppy/dpnp_glue/stubs.py [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+
+class dpnp(Stub):
+    """dpnp namespace
+    """
+    _description_ = '<dpnp>'
+
+    class sum(Stub):
+        pass
+
+    class eig(Stub):
```
Comment (diptorupd) [inline numba_dppy/dpnp_glue/stubs.py:11]: same here.

File: numba_dppy/dppy_passbuilder.py [hunk 14]
```diff
@@ -2,30 +2,49 @@
```
Comment (diptorupd) [inline numba_dppy/dppy_passbuilder.py:2]: Was this file reformatted using black? We should avoid mixing reformatting with logical changes. Reformatting should be a separate PR. Otherwise it creates too much noise.

File: numba_dppy/dppy_passes.py [hunk 15]
```diff
@@ -3,6 +3,7 @@
 import warnings
```
Comment (diptorupd) [inline numba_dppy/dppy_passes.py:3]: Does it make sense to organize all of our passes into a submodule?

File: numba_dppy/rename_numpy_functions_pass.py [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+
+
+class RewriteNumPyOverloadedFunctions(object):
+    def __init__(self, state, rewrite_function_name_map=rewrite_function_name_map):
+        self.state = state
+        self.function_name_map = rewrite_function_name_map
+
+    def run(self):
+        """
+        This function rewrites the name of NumPy functions that exist in self.function_name_map
```
Comment (diptorupd) [inline numba_dppy/rename_numpy_functions_pass.py:16]: Thanks for the nice docstring. We can run black for the new file. The comments do not conform tho PEP-8

File: numba_dppy/rename_numpy_functions_pass.py [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+
+        for label in topo_order:
+            block = blocks[label]
+            saved_arr_arg = {}
+            new_body = []
+            for stmt in block.body:
+                if isinstance(stmt, ir.Assign) and isinstance(stmt.value, ir.Expr):
+                    lhs = stmt.target.name
+                    rhs = stmt.value
+                    # replace np.func(sum) with name from self.function_name_map
```
Comment (diptorupd) [inline numba_dppy/rename_numpy_functions_pass.py:52]: not just `np.sum` right?

File: numba_dppy/rename_numpy_functions_pass.py [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+                            scope = global_module.scope
+                            loc = global_module.loc
+
+                            g_dppy_var = ir.Var(
+                                scope, mk_unique_var("$dppy_replaced_var"), loc
+                            )
+                            # We are trying to rename np.function_name/np.linalg.function_name with
+                            # numba_dppy.dpnp.function_name.
+                            # Hence, we need to have a global variable representing module numba_dppy.
+                            # Next, we add attribute dpnp to global module numba_dppy to
```
Comment (diptorupd) [inline numba_dppy/rename_numpy_functions_pass.py:80]: Thanks! the comment helps a lot.

File: numba_dppy/rename_numpy_functions_pass.py (hunk 16 above)
Comment (reazulhoque) [reply numba_dppy/rename_numpy_functions_pass.py:16]: Ran black 20.8b1 still did not reformat it. I manually made the line length smaller.

File: numba_dppy/dppy_passes.py (hunk 15 above)
Comment (reazulhoque) [reply numba_dppy/dppy_passes.py:3]: It does, created issue to track that. https://github.com/IntelPython/numba-dppy/issues/58

**Review by diptorupd — APPROVED** [review]
LGTM! merging this PR. Thanks!
